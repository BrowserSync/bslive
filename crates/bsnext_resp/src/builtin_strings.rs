use crate::connector::Connector;
use crate::debug::Debug;
use crate::injector_guard::ByteReplacer;
use crate::js_connector::JsConnector;
use axum::extract::Request;
use bsnext_guards::route_guard::RouteGuard;
use http::{Response, Uri};

#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub struct BuiltinStringDef {
    pub name: BuiltinStrings,
}
#[derive(Debug, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub enum BuiltinStrings {
    #[serde(rename = "bslive:connector")]
    Connector,
    #[serde(rename = "bslive:js-connector")]
    JsConnector,
    #[serde(rename = "bslive:debug")]
    Debug,
}

impl RouteGuard for BuiltinStringDef {
    fn accept_req(&self, req: &Request, outer_uri: &Uri) -> bool {
        match self.name {
            BuiltinStrings::Connector => Connector.accept_req(req, outer_uri),
            BuiltinStrings::Debug => Debug.accept_req(req, outer_uri),
            BuiltinStrings::JsConnector => JsConnector.accept_req(req, outer_uri),
        }
    }

    fn accept_res<T>(&self, res: &Response<T>, outer_uri: &Uri) -> bool {
        match self.name {
            BuiltinStrings::Connector => Connector.accept_res(res, outer_uri),
            BuiltinStrings::Debug => Debug.accept_res(res, outer_uri),
            BuiltinStrings::JsConnector => JsConnector.accept_res(res, outer_uri),
        }
    }
}

impl ByteReplacer for BuiltinStringDef {
    fn apply(&self, body: &'_ str) -> Option<String> {
        match self.name {
            BuiltinStrings::Connector => Connector.apply(body),
            BuiltinStrings::Debug => Debug.apply(body),
            BuiltinStrings::JsConnector => JsConnector.apply(body),
        }
    }
}
