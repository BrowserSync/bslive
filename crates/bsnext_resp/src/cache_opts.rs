use http::header::{CACHE_CONTROL, EXPIRES, PRAGMA};
use http::{HeaderName, HeaderValue};

#[derive(Debug, Default, PartialEq, Hash, Clone, serde::Deserialize, serde::Serialize)]
pub enum CacheOpts {
    /// Try to prevent browsers from caching the responses. This is the default behaviour
    #[serde(rename = "prevent")]
    #[default]
    Prevent,
    /// Do not try to prevent browsers from caching the responses.
    #[serde(rename = "default")]
    Default,
}

impl CacheOpts {
    pub fn as_headers(&self) -> Vec<(HeaderName, HeaderValue)> {
        match self {
            CacheOpts::Prevent => {
                vec![
                    (
                        CACHE_CONTROL,
                        HeaderValue::from_static("no-store, no-cache, must-revalidate"),
                    ),
                    (PRAGMA, HeaderValue::from_static("no-cache")),
                    (EXPIRES, HeaderValue::from_static("0")),
                ]
            }
            CacheOpts::Default => {
                vec![]
            }
        }
    }
}
