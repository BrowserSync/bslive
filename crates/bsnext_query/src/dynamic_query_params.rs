use axum::extract::Query;
use axum::response::IntoResponse;
use bsnext_resp::builtin_strings::BuiltinStrings;
use bsnext_resp::cache_opts::CacheOpts;
use http::Uri;

#[doc = include_str!("./query-params.md")]
#[derive(Debug, serde::Deserialize)]
pub struct DynamicQueryParams {
    /// Allow a request to have a ?bslive.delay.ms=200 style param to simulate a TTFB delay
    #[serde(rename = "bslive.delay.ms")]
    pub delay: Option<u64>,
    /// Control if Browsersync will add cache-busting headers, or not.
    #[serde(rename = "bslive.cache")]
    pub cache: Option<CacheOpts>,
    /// Control if Browsersync will add cache-busting headers, or not.
    #[serde(rename = "bslive.inject")]
    pub inject: Option<InjectParam>,
}

#[derive(Debug, serde::Deserialize)]
#[serde(untagged)]
pub enum InjectParam {
    BuiltinStrings(BuiltinStrings),
    Other(String),
}

#[cfg(test)]
mod test {
    use super::*;
    use axum::http::Uri;
    #[test]
    fn test_deserializing() -> anyhow::Result<()> {
        let input = Uri::from_static(
            "/abc?bslive.delay.ms=2000&bslive.cache=prevent&bslive.inject=bslive:js-connector",
        );
        let Query(query_with_named): Query<DynamicQueryParams> =
            Query::try_from_uri(&input).unwrap();
        insta::assert_debug_snapshot!(query_with_named);

        let input = Uri::from_static("/abc?bslive.inject=false");
        let Query(query_with_bool): Query<DynamicQueryParams> =
            Query::try_from_uri(&input).unwrap();
        insta::assert_debug_snapshot!(query_with_bool);
        Ok(())
    }
}
