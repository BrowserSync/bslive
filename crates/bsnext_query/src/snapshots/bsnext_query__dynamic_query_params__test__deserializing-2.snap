---
source: crates/bsnext_query/src/dynamic_query_params.rs
expression: query_with_bool
---
DynamicQueryParams {
    delay: None,
    cache: None,
    inject: Some(
        Other(
            "false",
        ),
    ),
}
